### PR TITLE
[prop-types] Do not consider null without undefined as optional

### PR DESCRIPTION
**Pull request by davidje13**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
  - some failing tests remain in `airbnb-prop-types`; should I fix those in the same PR?
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/37999
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

Authors: @DovydasNavickas @ferdaber @eps1lon

---

<s>This change works fine with my own code, though due to a bug (?) in the react types I haven't been able to run the formal test suite on my machine. "Error: Errors in typescript@next for external dependencies: ../react/index.d.ts(28,22): error TS2307: Cannot find module 'csstype'.". Hopefully the PR process will run tests on a CI machine without this issue.</s>

This both simplifies the `IsOptional` check and makes it behave better. Since the `Validator` type already includes `null | undefined` anyway, this won't make any difference to the built-in validators (they will still be considered optional unless marked as `isRequired`), but custom validators which consider `null` and `undefined` separately are now inferred correctly by `shape`.

<s>I would like to add a test for this use-case as well, but since I can't run the tests myself and I'm not familiar with this testing framework, I want the first commit to just ensure the existing tests do not break.</s>

### CODE REVIEW

**Comment by davidje13** [comment]
Hmm, the remaining test failures are strange. I'm wondering if this is a typescript issue, or something I don't understand in how these types are being inferred;

If I explicitly cast the optional type validators as `PropTypes.Validator<(whatever) | undefined | null>`, they work, whereas `PropTypes.Requireable<(whatever)>` fails (somehow the `undefined` possibility is being lost). But `Requireable<T> extends Validator<T | undefined | null>`

Nothing seems to look for `extends Requireable` explicitly, so the above 2 cases should be identical. Any ideas?

**Comment by davidje13** [comment]
OK so the tests are now all passing (apologies for the back-and-forth in the commit history; for the reasons explained in the first comment I haven't been able to run the tests properly on my local machine)

I had to switch to a `nominalTypeHack` style which didn't interfere with `undefined`. I notice that's an exported symbol; if anybody is using it from outside it will break, but should it really be exported?

I also added tests for nullable-only and undefinable-only validators, which was the original point here.

The remaining test failures are all in `airbnb-prop-types` which is making rather precise assertions on `Requireable<number | null>`, etc. which have now become `Requireable<number | null | undefined>`. Obviously the actual behaviour is unchanged, because `Requireable` has always added `null | undefined` to the type.

What's the usual procedure when changes are needed in a downstream project's tests? Do I just fix those tests in this PR too?

**Review by eps1lon — COMMENTED** [review]
New types look way friendlier. Thanks :+1: I'll let @ferdaber decide if this is ok since he's is the original code owner.

@dt-bot should mention downstream package collaborators. If they don't respond then it should be ok.

File: types/prop-types/prop-types-tests.ts
Comment (ferdaber) [inline types/prop-types/prop-types-tests.ts]: Why did this and `shape.bar` below it need to be changed to both be nullable?

**Comment by ferdaber** [comment]
The changes in two of the test suites' types concern me that this would be a breaking change. I remember that there were TS devs who were strongly opinionated that they be able to define their props interfaces more precisely, but still have propTypes for JS users (e.g. library authors). So they want to be able to keep their types a bit more precise for consumption within their own codebase, but have prop types for consumers outside their codebase, if that makes sense.

File: types/prop-types/prop-types-tests.ts
Comment (davidje13) [reply types/prop-types/prop-types-tests.ts]: To be honest I'm not sure why this wasn't needed before. It came up as an error so I happily patched it, because `| null` is the more correct type (`PropTypes.string` would allow `null` through). I understand that it may not be desirable from a practical standpoint, but maybe the better way to resolve that is to remove `|null` from the `Requireable` type? At least that way it's implemented in a consistent way (notice that the root-level optional type *already* had to include `null`; the change now is that nested types match the same behaviour)

**Comment by davidje13** [comment]
@ferdaber that makes sense; see my response to your inline comment above.

The problem with the old code is that it's inconsistent; the top-level `optionalNumber` already needed `| null`, so it's strange that the sub-properties `foo` and `bar` didn't, when they're defined the same way.

I don't know why that inconsistency happened in the old code (I can't see any logical reason for it in the types, and my intuition would be that `| null` would be needed).

There's no great way to relax this while also maintaining correctness for those users who want correctness. Presumably the most common issue would be types marked as optional where the library author doesn't want to bother considering the `null` case. That could be the source of a bug (if they guard against `undefined` explicitly, `null`s would slip through), but most of the time guards will be against falsey values, so it just adds noise to the type.

One option might be to introduce a `LaxNulls<T extends Validator>` or similar, which has the sole purpose of removing `| null` from all properties in `T`. Another option is to remove `| null` from `Requireable` (though that will negatively affect those who want correct types).

**Comment by ferdaber** [comment]
The most important functionality is to preserve type inference for JavaScript users, as all of these revolve around the `LibraryManagedAttributes` interface in the `JSX` namespace. As long as, for a JS user, the correct types are inferred via prop types, I'm fine with whatever changes— just be wary that you may have some angry TS users coming at you after this change goes through 😜

**Review by ferdaber — APPROVED** [review]

**Comment by davidje13** [comment]
Yes I expect that even this small change in a popular library like prop-types will generate some anger (sadly the semver strategy doesn't work for DefinitelyTyped packages)

It is change in the right direction though (i.e. encourages safer, more correct code); maybe somebody else can suggest a better way to allow lax null checking on an opt-in basis.

Also worth noting that this only affects those with StrictNullChecks enabled in tsconfig (which presumably biases towards developers who care about nullability anyway), and those who have nested properties (since top-level properties already did this).

**Comment by rbuckton** [comment]
Thanks for the contribution!

**Pull request merged**
